python/ray/data/_internal/stats.py

```diff
@@ -275,137 +332,302 @@ def summary_string(
             out += "Stage {} {}: executed in {}s\n".format(
                 self.number, self.base_name, rounded_total
             )
-            for n, (stage_name, metadata) in enumerate(self.stages.items()):
+            for n, stage_stats_summary in enumerate(self.stages_stats):
+                stage_name = stage_stats_summary.stage_name
                 stage_uuid = self.dataset_uuid + stage_name
                 out += "\n"
                 out += "\tSubstage {} {}: ".format(n, stage_name)
                 if stage_uuid in already_printed:
                     out += "\t[execution cached]\n"
                 else:
                     already_printed.add(stage_uuid)
-                    out += self._summarize_blocks(metadata, is_substage=True)
-        out += self._summarize_iter()
+                    out += str(stage_stats_summary)
+        out += str(self.iter_stats)
+        if self.extra_metrics:
+            indent = "\t" if stage_stats_summary.is_substage else ""
+            out += indent
+            out += "* Extra metrics: " + str(self.extra_metrics) + "\n"
         return out
 
-    def _summarize_iter(self) -> str:
-        out = ""
-        if (
-            self.iter_total_s.get()
-            or self.iter_wait_s.get()
-            or self.iter_next_batch_s.get()
-            or self.iter_format_batch_s.get()
-            or self.iter_get_s.get()
-        ):
-            out += "\nDataset iterator time breakdown:\n"
-            out += "* In ray.wait(): {}\n".format(fmt(self.iter_wait_s.get()))
-            out += "* In ray.get(): {}\n".format(fmt(self.iter_get_s.get()))
-            out += "* In next_batch(): {}\n".format(fmt(self.iter_next_batch_s.get()))
-            out += "* In format_batch(): {}\n".format(
-                fmt(self.iter_format_batch_s.get())
-            )
-            out += "* In user code: {}\n".format(fmt(self.iter_user_s.get()))
-            out += "* Total time: {}\n".format(fmt(self.iter_total_s.get()))
-        return out
+    def get_total_wall_time(self) -> float:
+        parent_sum = sum(p.get_total_wall_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.WALL_TIME).get("max", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_total_cpu_time(self) -> float:
+        parent_sum = sum(p.get_total_cpu_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.CPU_TIME).get("sum", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_max_heap_memory(self) -> float:
+        parent_memory = [p.get_max_heap_memory() for p in self.parents]
+        parent_max = max(parent_memory) if parent_memory else 0
+        return max(
+            parent_max,
+            *[
+                getattr(ss, StageStatsSummary.PEAK_HEAP_MEMORY).get("max", 0)
+                for ss in self.stages_stats
+            ],
+        )
+
+
+@dataclass
+class StageStatsSummary:
```

Comment:
Should this stats summary class (and the `IterStatsSummary` class) also have a `DeveloperAPI` designation? If a developer is wanting programmatic access to the `DatasetStatsSummary` class, they might want access to the stage stats and iter stats classes. cc @c21 @jianoaix, what do y'all think?

Reply:
I am thinking to not expose more APIs until we really need. Shall we add `DeveloperAPI` later? We can expose it after this becomes more stable, assuming some more change may apply to it as we work on observability.

Reply:
Yeah, the stage_stats and iter_stats fields of `DatasetStatsSummary` are public, so I think they are transitively DeveloperAPI (a developer gets a `DatasetStatsSummary` object can legitimately access its public fields).

Reply:
Sounds good!